File: gtci/03-sliding-window/cpp/6-longest-subarray-with-ones-after-replacement.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int findLength(const vector<int> &arr, int k) {
        int maxLength = 0;

        int windowStart = 0;
        int zeroCount = 0;
        for (int windowEnd = 0; windowEnd < arr.size(); windowEnd++) {
            int windowSize = windowEnd - windowStart + 1;

            if (arr[windowEnd] == 0) zeroCount++;

            if (zeroCount <= k) maxLength = std::max(maxLength, windowSize);

            if (zeroCount > k) {
                if (arr[windowStart] == 0) zeroCount--;

                windowStart++;
                windowSize = windowEnd - windowStart + 1;
            }
        }

        return maxLength;
    }
};

int main() {
    Solution sol;

    assert(sol.findLength({0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1}, 2) == 6);
    assert(sol.findLength({0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1}, 3) == 9);
    assert(sol.findLength({1, 0, 0, 1, 1, 0, 1, 1}, 2) == 6);

    std::cout << "All test cases passed." << std::endl;
    return 0;
}

File: gtci/03-sliding-window/python3/6-longest-subarray-with-ones-after-replacement.py
class Solution:
    def findLength(self, arr, k):
        max_length = 0

        zero_count = 0
        window_start = 0
        for window_end in range(len(arr)):
            window_size = window_end - window_start + 1

            if arr[window_end] == 0:
                zero_count += 1

            if zero_count <= k:
                max_length = max(max_length, window_size)

            while zero_count > k:
                if arr[window_start] == 0:
                    zero_count -= 1

                window_start += 1
                window_size = window_end - window_start + 1

        return max_length


if __name__ == "__main__":
    sol = Solution()
    assert sol.findLength([0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1], 2) == 6
    assert sol.findLength([0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1], 3) == 9
    assert sol.findLength([1, 0, 0, 1, 1, 0, 1, 1], 2) == 6

    print("All test cases passed.")
